<a href="https://colab.research.google.com/github/Ganesh-Esc/Deep-Learning---Reinforcement-Learning/blob/main/Reinforcement_Learning_DEMO.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##  Reinforcement Learning DEMO

## Author

<a href="https://www.linkedin.com/in/ganesh-kommana/" target="_blank">Ganesh Kommana</a>

## Reinforcement Learning Example

In this example from Reinforcement Learning, the task is to use tools from Machine Learning to predict how an agent should act. We will then use those predictions to drive the behavior of the agent. Ideally, our intelligent agent should get a much better score than a random agent.

## Key concepts:

- **Observation**: These are the states of the game. It describes where the agent currently is.
- **Action**: These are the moves that the agent makes.
- **Episode**: One full game played from beginning (`env.reset()`) to end (when `done == True`).
- **Step**: Part of a game that includes one action. The game transitions from one observation to the next.

## Setup

This exaple uses the Python library [OpenAI Gym](https://gym.openai.com/docs/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkML311Coursera747-2022-01-01).

If you want to install everything (gym can run atari games.) follow [these instructions](https://github.com/openai/gym#installing-everything).

Now we can build an environment using OpenAI.

In [1]:
!mamba install -qy -c conda-forge gym==0.21.0

/bin/bash: line 1: mamba: command not found


In [17]:
import gym
import pandas
import numpy as np

import warnings
warnings.filterwarnings("ignore")

def warn(*args, **kwargs):
    return None

warnings.warn = warn

# The first part of the game uses the environment FrozenLake-V0

This is a small world with 16 tiles.

    SFFF
    FHFH
    FFFH
    HFFG

The game starts at the S tile. The object of the game is to get to the goal (G) without landing in a hole (H).


In [18]:
# Build an environment with gym.make()
env = gym.make('FrozenLake-v1') # Build a fresh environment

# Start a new game with env.reset()
current_observation = env.reset() # This starts a new "episode" and returns the initial observation

# The current observation is just the current location
print(current_observation) # Observations are just a number

0


In [19]:
# we can print the environment if we want to look at it
env.render()

In [20]:
# the action space for this environment includes four discrete actions

print(f"our action space: {env.action_space}")

new_action = env.action_space.sample() # we can randomly sample actions

print(f"our new action: {new_action}") # run this cell a few times to get an idea of the action space
# what does it look like?

our action space: Discrete(4)
our new action: 3


In [22]:
# now we act! do this with the step function

new_action = env.action_space.sample()

observation, reward, done, info = env.step(new_action)

# here's a look at what we get back
print(f"observation: {observation}, reward: {reward}, done: {done}, info: {info}")

env.render()

observation: 0, reward: 0.0, terminated: False, truncated: False, info: {'prob': 0.3333333333333333}


In [23]:
!pip install numpy==1.24.3

  Using cached numpy-1.24.3.tar.gz (10.9 MB)
  Installing build dependencies ... done
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Getting requirements to build wheel ... error
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.


In [27]:
# we can put this process into a for-loop and see how the game progresses

current_observation, info = env.reset() # start a new game

for i in range(5): # run 5 moves

    new_action = env.action_space.sample() # same a new action

    observation, reward, terminated, truncated, info = env.step(new_action) # step through the action and get the outputs

    # here's a look at what we get back
    print(f"observation: {observation}, reward: {reward}, terminated: {terminated}, truncated: {truncated}, info: {info}")

    env.render()

observation: 0, reward: 0.0, terminated: False, truncated: False, info: {'prob': 0.3333333333333333}
observation: 0, reward: 0.0, terminated: False, truncated: False, info: {'prob': 0.3333333333333333}
observation: 0, reward: 0.0, terminated: False, truncated: False, info: {'prob': 0.3333333333333333}
observation: 0, reward: 0.0, terminated: False, truncated: False, info: {'prob': 0.3333333333333333}
observation: 0, reward: 0.0, terminated: False, truncated: False, info: {'prob': 0.3333333333333333}


Now we can guess what each of the outputs mean.

**Observation** refers to the number of the tile. The tiles appear to be numbered

    0 1 2 3
    4 5 ...
    
**Reward** refers to the outcome of the game. We get 1 if we win, zero otherwise.

**Done** tells us if the game is still going. It goes to true when we win or fall into a hole.

**info** gives extra info about the world. Here, it's probabilities. Can you guess what this means here? Perhaps the world is a bit noisy.

In [28]:
# Here's how to simulate an entire episode
# We're going to stop rendering it every time to save space
# try running this a few. Does it ever win?

current_observation = env.reset()
done = False

while not done:
    new_action = env.action_space.sample()
    new_observation, reward, done, info = env.step(new_action)
    print(f"action:{new_action} observation: {new_observation}, reward: {reward}, done: {done}, info: {info}")

action:1 observation: 4, reward: 0.0, terminated: False, truncated: False, info: {'prob': 0.3333333333333333}
action:2 observation: 0, reward: 0.0, terminated: False, truncated: False, info: {'prob': 0.3333333333333333}
action:0 observation: 0, reward: 0.0, terminated: False, truncated: False, info: {'prob': 0.3333333333333333}
action:1 observation: 4, reward: 0.0, terminated: False, truncated: False, info: {'prob': 0.3333333333333333}
action:3 observation: 5, reward: 0.0, terminated: True, truncated: False, info: {'prob': 0.3333333333333333}


Things to think about:
- What things do you notice about how the environment and actions work?
- What do you think the actions mean?
- When the agent performs the same action from the same place (same observation), does the same outcome happen every time?

The environment has some squares that always end the game (`H` in the render), some that don't (`F`), and one that is presumably the reward, if you get to it.

The actions seem like up, down, left, and right. But they also seem stochastic. There seems to be a 1/3 chance of going into 3 different squares with each action.


# Part 1: Gather data

We want to build an intelligent actor but first we have to gather data on which actions are useful.

Use the code above as reference. Run a *random* agent through 1,000 or more episodes and collect data on each step.

I recommend you store this data in a pandas dataframe. Each row should be a step. Your features should include the following features or similar

- `observation` the observation at the beginning of the step (before acting!)
- `action` the action randomly sampled
- `current_reward` the reward received after the action was performed

After you generate this data, it is recommended that you compute a column (such as `total_reward`, that is the total reward for the entire episode).

At the end of the data gathering, you should be able to use pandas (or similar) to calculate the average total reward *per episode* of the random agent. The average score should be 1-2%, meaning that the agent very rarely wins.


## Hints

- `initial_observation = env.reset()` starts a new episode and returns the initial observation.
- `new_observation, reward, done, info = env.step(new_action)` executes one action and returns the following observation. You may look at the documentation for the step method if you are curious about what it does.
- `done != True` until the game is finished.
- we are trying to maximize the reward *per episode*. Our first game gives 0 reward unless the agent travels to the goal.
- `env.action_space.n` gives the number of possible actions in the environment. `env.action_space.sample()` allows the agent to randomly sample an action.
- `env.observation_space.n` gives the number of possible states in the environment.

In [30]:
env = gym.make('FrozenLake-v1')

num_episodes = 40000

life_memory = []
for i in range(num_episodes):

    # start a new episode and record all the memories
    old_observation = env.reset()
    done = False
    tot_reward = 0
    ep_memory = []
    while not done:
        new_action = env.action_space.sample()
        observation, reward, done, info = env.step(new_action)
        tot_reward += reward

        ep_memory.append({
            "observation": old_observation,
            "action": new_action,
            "reward": reward,
            "episode": i,
        })
        old_observation = observation

    # incorporate total reward
    num_steps = len(ep_memory)
    for i, ep_mem in enumerate(ep_memory):
        ep_mem["tot_reward"] = tot_reward
        ep_mem["decay_reward"] = i*tot_reward/num_steps

    life_memory.extend(ep_memory)

memory_df = pandas.DataFrame(life_memory)

In [12]:
!mamba install -qy -c conda-forge numpy=1.24.3

/bin/bash: line 1: mamba: command not found


In [31]:
memory_df.describe()

,observation,action,reward,episode,tot_reward,decay_reward
count,308869.000000,308869.000000,308869.000000,308869.000000,308869.000000,308869.000000
mean,2.226258,1.500895,0.001871,20015.807718,0.025247,0.011688
std,3.002619,1.118500,0.043219,11529.689415,0.156875,0.085856
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,0.000000,0.000000,10060.000000,0.000000,0.000000
50%,1.000000,2.000000,0.000000,20026.000000,0.000000,0.000000
75%,4.000000,3.000000,0.000000,29968.000000,0.000000,0.000000
max,14.000000,3.000000,1.000000,39999.000000,1.000000,0.979592


In [32]:
memory_df.shape

(308869, 6)

In [33]:
memory_df.groupby("episode").reward.sum().mean()

np.float64(0.01445)

# Step 2: Predict

Now that you have a bunch of data put it into a format that you can model. The goal here is to guide the behavior of our agent. Our agent will be given an observation and need to decide between the possible actions given that observation and the prediction of the model.

Remember, you're a data scientist! Be creative.

It might be helpful to work backwards. Ultimately, you will write something like:

```
def convert_to_row(obs, act):
    # expertly written code
    return row_of_obs_act
    
rows = [convert_to_row(current_obs, act) for act in possible_actions]

pred_outcome = model.predict(rows)
```

So, you will need to design a quantity that you can ask your model to predict for every possible action-observation pair. Think a bit about what this quantity should be. Should the model try to predict the immediate reward for each action? If so, how would it know where to go at the beginning of each episode when all moves give zero reward but when some moves bring it closer to the goal than others.


# Step 2: Predict

Now that you have a bunch of data put it into a format that you can model. The goal here is to guide the behavior of our agent. Our agent will be given an observation and need to decide between the possible actions given that observation and the prediction of the model.

Remember, you're a data scientist! Be creative.

It might be helpful to work backwards. Ultimately, you will write something like:

```
def convert_to_row(obs, act):
    # expertly written code
    return row_of_obs_act
    
rows = [convert_to_row(current_obs, act) for act in possible_actions]

pred_outcome = model.predict(rows)
```

So, you will need to design a quantity that you can ask your model to predict for every possible action-observation pair. Think a bit about what this quantity should be. Should the model try to predict the immediate reward for each action? If so, how would it know where to go at the beginning of each episode when all moves give zero reward but when some moves bring it closer to the goal than others.


In [34]:
from sklearn.ensemble import RandomForestRegressor, ExtraTreesRegressor
from sklearn.svm import SVR

model = ExtraTreesRegressor(n_estimators=50)
# model = SVR()
y = 0.5*memory_df.reward + 0.1*memory_df.decay_reward + memory_df.tot_reward
x = memory_df[["observation", "action"]]
model.fit(x, y)

ExtraTreesRegressor(n_estimators=50)

# Step 3: Act

Now that you have a model that predicts the desired behavior, let's act on it! Modify the code you used to gather data so that you replace the random decision with an intelligent one.

We started out winning ~1.5% of the games with the random agent. How well can you do? You should be able to get your model to do at least 10x better (so 15%). Can you get ~50%?

If you're having trouble, tune your model. Try different representations of the observation and action spaces. Try different models.


In [36]:
model = RandomForestRegressor()
y = 1*memory_df.reward + memory_df.tot_reward + .1*memory_df.decay_reward
x = memory_df[["observation", "action"]]
model.fit(x, y)

num_episodes = 500
random_per = 0

life_memory = []
for i in range(num_episodes):
    # Start a new episode and record all the memories.
    old_observation, info = env.reset() # Updated to unpack info as well
    terminated = False # Updated variable name
    truncated = False # Added truncated variable
    tot_reward = 0
    ep_memory = []
    while not (terminated or truncated): # Updated loop condition
        if np.random.rand() < random_per:
            new_action = env.action_space.sample()
        else:
            pred_in = [[old_observation,i] for i in range(4)]
            new_action = np.argmax(model.predict(np.array(pred_in))) # Converted pred_in to NumPy array
        observation, reward, terminated, truncated, info = env.step(new_action) # Updated to unpack 5 values
        tot_reward += reward

        ep_memory.append({
            "observation": old_observation,
            "action": new_action,
            "reward": reward,
            "episode": i,
        })
        old_observation = observation

    # incorporate total reward
    for ep_mem in ep_memory:
        ep_mem["tot_reward"] = tot_reward

    life_memory.extend(ep_memory)

memory_df2 = pandas.DataFrame(life_memory)

# rf.fit(memory_df[["observation", "action"]], memory_df["comb_reward"])

# Score
# Much better!
memory_df2.groupby("episode").reward.sum().mean()

np.float64(0.748)

## Conclusion

We have successfully implemented an intelligent agent using a Random Forest Regressor to predict the potential reward for each action in the FrozenLake environment. By using the model's predictions to guide the agent's actions, we were able to significantly improve its performance compared to a random agent. The intelligent agent achieved an average reward of {{memory_df2.groupby("episode").reward.sum().mean()}}, which is much higher than the random agent's average reward of {{memory_df.groupby("episode").reward.sum().mean()}}.

In [37]:
memory_df.describe()

,observation,action,reward,episode,tot_reward,decay_reward
count,308869.000000,308869.000000,308869.000000,308869.000000,308869.000000,308869.000000
mean,2.226258,1.500895,0.001871,20015.807718,0.025247,0.011688
std,3.002619,1.118500,0.043219,11529.689415,0.156875,0.085856
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,0.000000,0.000000,10060.000000,0.000000,0.000000
50%,1.000000,2.000000,0.000000,20026.000000,0.000000,0.000000
75%,4.000000,3.000000,0.000000,29968.000000,0.000000,0.000000
max,14.000000,3.000000,1.000000,39999.000000,1.000000,0.979592


In [39]:
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor, ExtraTreesRegressor

model = ExtraTreesRegressor(n_estimators=50)

memory_df["comb_reward"] = .5*memory_df.reward + memory_df.tot_reward
model.fit(memory_df[["observation", "action"]], memory_df.comb_reward)

ExtraTreesRegressor(n_estimators=50)

In [41]:
num_episodes = 100
random_per = 0

life_memory = []
for i in range(num_episodes):

    # start a new episode and record all the memories
    old_observation_tuple = env.reset()
    old_observation = old_observation_tuple[0] # Extract the integer observation
    terminated = False
    truncated = False
    tot_reward = 0
    ep_memory = []
    while not (terminated or truncated):


        if np.random.rand() < random_per:
            new_action = env.action_space.sample()
        else:
            pred_in = [[old_observation, action_val] for action_val in range(env.action_space.n)] # Create input for all possible actions
            new_action = np.argmax(model.predict(np.array(pred_in))) # Convert to numpy array and predict
        observation, reward, terminated, truncated, info = env.step(new_action) # Unpack 5 values
        tot_reward += reward

        ep_memory.append({
            "observation": old_observation, # Use the single observation value
            "action": new_action,
            "reward": reward,
            "episode": i,
        })
        old_observation = observation

    # incorporate total reward
    for ep_mem in ep_memory:
        ep_mem["tot_reward"] = tot_reward

    life_memory.extend(ep_memory)

memory_df2 = pandas.DataFrame(life_memory)
memory_df2["comb_reward"] = memory_df2.reward + memory_df2.tot_reward

# score
# much better!
memory_df2.groupby("episode").reward.sum().mean()

np.float64(0.85)

The conclusion from this cell's execution is that the intelligent agent, using the trained model, achieved an average reward of 0.85. This is a substantial improvement compared to the random agent, which had an average reward of approximately 0.01445 (as calculated in cell 28TCWWGcFmVp). This demonstrates that the model was effective in learning a policy that allows the agent to navigate the environment and reach the goal more consistently.